cmd/maya-apiserver/cstor-operator/cspc/controller.go

```diff
@@ -0,0 +1,208 @@
+	cb.Controller.cspcSynced = cspcInformer.Informer().HasSynced
+	return cb
+}
+
+// withWorkqueue adds workqueue to controller object.
+func (cb *ControllerBuilder) withWorkqueueRateLimiting() *ControllerBuilder {
+	cb.Controller.workqueue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CSPC")
+	return cb
+}
+
+// withRecorder adds recorder to controller object.
+func (cb *ControllerBuilder) withRecorder(ks kubernetes.Interface) *ControllerBuilder {
+	glog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: ks.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+	cb.Controller.recorder = recorder
+	return cb
+}
+
+// withEventHandler adds event handlers controller object.
+func (cb *ControllerBuilder) withEventHandler(cspcInformerFactory informers.SharedInformerFactory) *ControllerBuilder {
+	cspcInformer := cspcInformerFactory.Openebs().V1alpha1().CStorPoolClusters()
+	// Set up an event handler for when CSPC resources change
+	cspcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    cb.Controller.addSpc,
+		UpdateFunc: cb.Controller.updateSpc,
+		// This will enter the sync loop and no-op, because the cspc has been deleted from the store.
+		DeleteFunc: cb.Controller.deleteSpc,
+	})
+	return cb
+}
+
+// Build returns a controller instance.
+func (cb *ControllerBuilder) Build() (*Controller, error) {
+	err := openebsScheme.AddToScheme(scheme.Scheme)
+	if err != nil {
+		return nil, err
+	}
+	return cb.Controller, nil
+}
+
+// addSpc is the add event handler for cspc
+func (c *Controller) addSpc(obj interface{}) {
+	cspc, ok := obj.(*apis.CStorPoolCluster)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("Couldn't get cspc object %#v", obj))
+		return
+	}
+	if cspc.Annotations[string(apis.OpenEBSDisableReconcileKey)] == "true" {
+		message := fmt.Sprintf("reconcile is disabled via %q annotation", string(apis.OpenEBSDisableReconcileKey))
+		c.recorder.Event(cspc, corev1.EventTypeWarning, "Create", message)
+		return
+	}
+	glog.V(4).Infof("Queuing CSPC %s for add event", cspc.Name)
+	c.enqueueSpc(cspc)
+}
+
+// updateSpc is the update event handler for cspc.
+func (c *Controller) updateSpc(oldSpc, newSpc interface{}) {
+	cspc, ok := newSpc.(*apis.CStorPoolCluster)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("Couldn't get cspc object %#v", newSpc))
+		return
+	}
+	if cspc.Annotations[string(apis.OpenEBSDisableReconcileKey)] == "true" {
+		message := fmt.Sprintf("reconcile is disabled via %q annotation", string(apis.OpenEBSDisableReconcileKey))
+		c.recorder.Event(cspc, corev1.EventTypeWarning, "Update", message)
+		return
+	}
+	// Enqueue cspc only when there is a pending pool to be created.
+	if c.isPoolPending(cspc) {
```

Comment:
this fn being doing `k8s api` is better to move out from k8s context (updateSpc is called from k8s context).. so just enqueue and let handler take care of it

Reply:
I was also in dilemma where to put it but then I gave little bit of thought and went through following and preferred the above way, although open to discuss and incorporate.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L255

Reply:
lets not make k8s calls from k8s context which can lead to recursion